### DISCUSSION TITLE
SQLAlchemy 2 /w dataclasses: Inherited columns in inherited tables not working?

### DISCUSSION THREAD

**Discussion by daemonkeeper** [sqlalchemy/sqlalchemy#8819] [Usage Questions] [answered]
Hi, I'm trying to figure out how to use inherited columns, in inherited classes. I'm using SQLAlchemy 2 Beta and dataclasses as a declarative base. Naively I tried

```python
class BaseModel(MappedAsDataclass, DeclarativeBase, kw_only=True):
    uuid: Mapped[UUID] = mapped_column(primary_key=True, default_factory=uuid4, repr=True)


class NotinheritedTable(BaseModel):
    __tablename__ = "notinherited"
    pass
```

This works perfectly fine for regular table models:

```python
>> normalclass = NotinheritedTable()
>> print(normalclass)
NotinheritedTable(uuid=UUID('f27b4ccd-f000-4912-abad-e3ecf0d267be'))
```

However, this stops working as soon as I have inherited classes (single-table inheritance; cf. https://docs.sqlalchemy.org/en/20/orm/inheritance.html#single-table-inheritance) as defined below

```python
class BaseModel(MappedAsDataclass, DeclarativeBase, kw_only=True):
    uuid: Mapped[UUID] = mapped_column(primary_key=True, default_factory=uuid4, repr=True)


class PrimaryKeyMixIn:
    #uuid: Mapped[UUID] = mapped_column(primary_key=True, default_factory=uuid4, repr=True)
    @declared_attr
    def uuid(cls): Mapped[UUID]:
        return cls.__table__.c.get(
            "uuid", mapped_column(primary_key=True, default_factory=uuid4, repr=True)
        )

class NotinheritedTable(BaseModel):
    __tablename__ = "notinherited"
    pass

class Employee(BaseModel):
    __tablename__ = "employee"
    base_column: Mapped[str] = mapped_column(init=True)
    table_type: Mapped[str] = mapped_column(init=False)

    __mapper_args__ = {
        "polymorphic_identity": "employee",
        "polymorphic_on": "table_type",
    }

class Manager(PrimaryKeyMixIn, Employee):
    __mapper_args__ = {
        "polymorphic_identity": "manager",
    }

class Engineer(Employee):
    subclass_column: Mapped[Optional[str]] = mapped_column(init=True)

    __mapper_args__ = {
        "polymorphic_identity": "engineer",
    }
```

According to the docs, I should be using mix-ins as the documentation suggests (https://docs.sqlalchemy.org/en/20/orm/declarative_mixins.html). I tried one sub-class with a mix-in, one without one. 

The Manager table with the mix-in gives:

```python
>> manager = Manager(base_column="manager table")
>> print(manager)
Manager(uuid=Column('uuid', Uuid(), table=<employee>, primary_key=True, nullable=False), base_column='manager table', table_type='manager')
```

The Engineer table

```python
>> engineer = Engineer(subclass_column="engineer", base_column="engineer table")
>> print(engineer)
Engineer(uuid=<sqlalchemy.orm.attributes.InstrumentedAttribute object at 0x1070e13a0>, base_column='engineer table', table_type='engineer', subclass_column='engineer')
```

Trying to actually commit this yields in both cases:

```
2022-11-15 15:37:18,698 INFO sqlalchemy.engine.Engine INSERT INTO employee (base_column, table_type, uuid, subclass_column) VALUES (?, ?, employee.uuid, ?) RETURNING uuid
2022-11-15 15:37:18,698 INFO sqlalchemy.engine.Engine [generated in 0.00004s] ('engineer table', 'engineer', 'engineer')
2022-11-15 15:37:18,698 INFO sqlalchemy.engine.Engine ROLLBACK
Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1942, in _exec_single_context
    self.dialect.do_execute(
  File "/opt/homebrew/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 743, in do_execute
    cursor.execute(statement, parameters)
sqlite3.OperationalError: no such column: employee.uuid
```

In Gitter, someone suggested to use the MappedAsDataclass mix-in in the inherited tables too, but that does not seem to make a difference, i.e.;

```python
class Manager(MappedAsDataclass, Employee):
    __mapper_args__ = {
        "polymorphic_identity": "manager",
    }
```

However, that only works in a certain order following Python inheritance rules, which may negate the mix-in anyway. 

I am not sure if I am doing anything wrong, as I understand the docs this should be precisely as it works, or if there may be a bug in the beta? Can anyone maybe shed some light into this? 

You can also check out the full file in my gists: https://gist.github.com/daemonkeeper/52b09580fce939243762469b02a13a4c

**Comment by zzzeek** [comment]
yeah I saw this on gitter and as the MappedAsDataclass thing is completely new, it's likely there's a bug of some kind here.  the general idea is that the dataclass definition is getting the whole "mapped()" structure as the default value for the `__init__()` method, and not your `default_factory`, hence when you make a new object, the raw `Column` ends up going into the object's state and you see that strange SQL result.

Will confirm this is on our end (the other possibility being, dataclasses are weird so perhaps there's some reason we can't get around this, but seems unlikely) and open a new issue.

**Comment by zzzeek** [comment, marked as answer]
OK this can't work as written because your `uuid()` attr is returning a raw `Column` object, which is not a `mapped_column()` object.  This is normally fine but when using dataclass mappings, all attributes have to be given as a `mapped_column()`.     I see this is on us because the docs at https://docs.sqlalchemy.org/en/20/orm/inheritance.html#resolving-column-conflicts are showing the pattern that you're using, and I dont know that we have a construct in place for this use case right now, so that would need to be added and documented in that section.


the next thing that's not lining up here is that the mapping has the "uuid" column on the base class already, so this is not the usual "single inheritance mixin" pattern, but in any case that shouldn't matter here.

so the pattern that should be used, which I will add some kind of API feature to address more succinctly, is for the moment:

```py
class PrimaryKeyMixIn:
    @declared_attr.cascading
    def uuid(cls) -> Mapped[UUID]:
        mc = mapped_column(
            primary_key=True, default_factory=uuid4, repr=True
        )
        existing_col = cls.__table__.c.get('uuid', None)
        if existing_col is not None:
            mc.column = existing_col
        return mc

```

since we have `mapped_column()` now that can do anything we need for the ORM, I will try, though I'm not having much success yet, of making this like:

```py
class PrimaryKeyMixIn:
    uuid: Mapped[UUID] = mapped_column(
        primary_key=True, default_factory=uuid4, repr=True,
        use_existing_column=True
    )
```

**Comment by daemonkeeper** [comment]
Thanks a lot, this appears to work. I seem to recall I did try something similar already, though without success. With that being said, this works as a workaround but the mapped_column argument you're proposing is much nicer. I assume, if deemed necessary, you'll open an issue about this, and the documentation bug you mentioned?

**Reply by zzzeek** [reply]
yes, GH doesn't note new issues here so it is #8822.
